### PR TITLE
Add uh court case migrator

### PR DESCRIPTION
**Pull request by antonyoneill**
## Context

We want to model court cases within MA. We're going to migrate the data in from UH on this so we can look up for formal agreements and allow officers to manage court cases through MA

## Changes proposed in this pull request

Initial start on wiring up a court case migrator. It's responsibility is to determine whether it can/should create a court case in MA, and do so.

It will update court cases on subsequent syncs, provided there's only a single court case, and that it won't overwrite existing data on that single court case.

## Link to Jira card

https://hackney.atlassian.net/browse/MAAP-417

## Things to check

- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] Environment variables have been updated

### CODE REVIEW

File: lib/hackney/income/migrate_uh_court_case.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
+module Hackney
+  module Income
+    class MigrateUhCourtCase
+      def initialize(create_court_case:, view_court_cases:, update_court_case:)
+        @create_court_case = create_court_case
+        @view_court_cases = view_court_cases
+        @update_court_case = update_court_case
+      end
+
+      def migrate(criteria)
```
Comment (elena-vi) [inline lib/hackney/income/migrate_uh_court_case.rb:10]: is criteria the best variable name..?

**Review by elena-vi — APPROVED** [review]
this looks good.

I'm just wondering if we could classify any tenancies that have no court date but a court outcome 
as `check case data`, I think that will help the users to clean up the data

File: spec/lib/hackney/income/migrate_uh_court_case_spec.rb [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+          )
+        )
+        subject
+      end
+    end
+
+    context 'when provided a criteria with a court outcome but without a court date' do
+      let(:criteria_attributes) {
+        {
+          court_outcome: Hackney::Tenancy::UpdatedCourtOutcomeCodes::ADJOURNED_FOR_DIRECTIONS_HEARING,
```
Comment (C-gyorfi) [inline spec/lib/hackney/income/migrate_uh_court_case_spec.rb:61]: I'm sure you are aware of this and probably we should address this in another slice: UH court cases are in: `Hackney::Tenancy::CourtOutcomeCodes`  and we will need to map these when migrating.

File: spec/lib/hackney/income/migrate_uh_court_case_spec.rb [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+          subject
+        end
+      end
+    end
+
+    context 'when a partial court case already exists without a court date but has an outcome' do
+      let(:existing_court_cases) {
+        [create(:court_case,
+                court_date: nil,
+                court_outcome: Hackney::Tenancy::CourtOutcomeCodes::ADJOURNED_GENERALLY)]
```
Comment (C-gyorfi) [inline spec/lib/hackney/income/migrate_uh_court_case_spec.rb:209]: Maybe I was wrong in my previous comment, are the `Hackney::Tenancy::CourtOutcomeCodes` whet we meant to use in MA court_cases?

**Review by C-gyorfi — APPROVED** [review]
Nice work, I'm just a bit confused about the court outcomes

File: lib/hackney/income/migrate_uh_court_case.rb (hunk 1 above)
Comment (antonyoneill) [reply lib/hackney/income/migrate_uh_court_case.rb:10]: Good question! For the purposes of this, it's a legacy data record... That said we do get the criteria and call it that throughout the code base, and we use a `StubCriteria`. So I think this is best for now.

File: spec/lib/hackney/income/migrate_uh_court_case_spec.rb (hunk 2 above)
Comment (antonyoneill) [reply spec/lib/hackney/income/migrate_uh_court_case_spec.rb:61]: ! Thanks so much. I forgot about mapping them! :disappointed:

File: spec/lib/hackney/income/migrate_uh_court_case_spec.rb (hunk 3 above)
Comment (antonyoneill) [reply spec/lib/hackney/income/migrate_uh_court_case_spec.rb:209]: I actually got a bit confused, there's two:
- `Hackney::Tenancy::CourtOutcomeCodes`
- `Hackney::Tenancy::UpdatedCourtOutcomeCodes`

I wasn't really sure what it meant..

**Comment by antonyoneill** [comment]
@elena-vi wrt to the check case data, we do already have that pointing to the old UH data. We should update/create a new MA ruleset like Csaba did for the agreement breach rulesets

File: lib/hackney/income/migrate_uh_court_case.rb (hunk 1 above)
Comment (elena-vi) [reply lib/hackney/income/migrate_uh_court_case.rb:10]: I like the idea of phasing out the term `criteria` on the leasehold side I've started calling the thing that we get from UH as [`attributes`](https://github.com/LBHackney-IT/lbh-income-api/blob/master/lib/hackney/leasehold/sync_action_attributes.rb#L10)

File: spec/lib/hackney/income/migrate_uh_court_case_spec.rb [hunk 4]
```diff
@@ -21,12 +21,15 @@
   let(:criteria) { Stubs::StubCriteria.new(criteria_attributes) }
   let(:existing_court_cases) { [] }
 
+  UH_NIL_COURT_OUTCOME = '   '.freeze
+  UH_NIL_DATE = DateTime.parse('1900-01-01 00:00:00')
```
Comment (C-gyorfi) [inline spec/lib/hackney/income/migrate_uh_court_case_spec.rb:25]: Ugh, maybe a lame question but why though 🤔

Comment (antonyoneill) [reply spec/lib/hackney/income/migrate_uh_court_case_spec.rb:25]: Why does UH do this? 
- It was fairly normal "back in the old days" to reserve the data space that you would potentially occupy for performance reasons. ie the `u_court_outcome` is defined as a 3 character varchar, so they'd pad it with spaces.. :eyes: 
- It's also likely that the form in UHW that they'd input the 3 characters would be prepopulated with spaces, and you'd replace over it - think that this system was probably designed and built by people that used to design the old terminal applications like
  image

Why do I have to do this?
- I don't want to create court cases in MA with the date at the beginning of the century, with a court outcome of `"   "` :joy:

Comment (C-gyorfi) [reply spec/lib/hackney/income/migrate_uh_court_case_spec.rb:25]: Thanks for explaining 🤓 👍

**Pull request merged**